### PR TITLE
use mio for Tcp management in bootstrap system

### PR DESCRIPTION
**Pull request by modship**
* [ ] document all added functions
* [x] try in sandbox /simulation/labnet
* [x] unit tests on the added/changed features
  * [x] make tests compile
  * [x] make tests pass 
* [ ] add logs allowing easy debugging in case the changes caused problems
* [ ] if the API has changed, update the API specification

### CODE REVIEW

File: massa-bootstrap/src/listener.rs [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+            .name("bs_listener".to_string())
+            .spawn(move || loop {
+                poll.poll(&mut events, None).unwrap();
+
+                for event in events.iter() {
+                    match event.token() {
+                        NEW_CONNECTION => {
+                            let (socket, addr) =
+                                server.accept().map_err(BootstrapError::from).unwrap();
+                            println!("New connection: {}", addr);
```
Comment (modship) [inline massa-bootstrap/src/listener.rs:47]: this println will be deleted

**Comment by modship** [comment]
@Ben-PH  i think `etablisher.rs` can be deleted

**Comment by Ben-PH** [comment]
@modship [d25f902](https://github.com/massalabs/massa/pull/3780/commits/d25f9020f78ff45f2b1e10d1904494f07ebf5344) should fix the event-loop breakout being blocked.

File: massa-bootstrap/src/listener.rs [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+pub enum PollEvent {
+    NewConnection((TcpStream, SocketAddr)),
+    Stop,
+}
+impl BootstrapTcpListener {
+    /// Start a new bootstrap listener on the given address.
+    ///
+    /// * `addr` - the address to listen on
+    /// * `connection_tx` - the channel to send new connections to
```
Comment (Ben-PH) [inline massa-bootstrap/src/listener.rs:34]:
```suggestion
    /// Setup a mio-listener that functions as a `select!` on a connection, or a waker
    ///
    /// * `addr` - the address to listen on
```

File: massa-bootstrap/src/listener.rs [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+use tracing::info;
+
+use crate::error::BootstrapError;
+use crate::establisher::BSEventPoller;
+
+const NEW_CONNECTION: Token = Token(0);
+const STOP_LISTENER: Token = Token(10);
+
+/// TODO: this should be crate-private. currently needed for models testing
+pub struct BootstrapTcpListener {
```
Comment (Ben-PH) [inline massa-bootstrap/src/listener.rs:15]: add follow up issue

File: massa-bootstrap/src/listener.rs [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+    poll: Poll,
+    events: Events,
+    server: TcpListener,
+    // HACK : create variable to move ownership of mio_server to the thread
+    // if mio_server is not moved, poll does not receive any event from listener
+    _mio_server: MioTcpListener,
+}
+
+pub struct BootstrapListenerStopHandle(Waker);
```
Comment (Ben-PH) [inline massa-bootstrap/src/listener.rs:24]: add follow up issue to make private/crate private

File: massa-bootstrap/src/listener.rs [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+pub enum PollEvent {
```
Comment (Ben-PH) [inline massa-bootstrap/src/listener.rs:26]: add follow-up issue to make private/crate private

File: massa-bootstrap/src/server.rs [hunk 6]
```diff
@@ -59,33 +59,56 @@ use white_black_list::*;
...
 /// over to the worker via a channel
-type BsConn = (TcpStream, SocketAddr);
 
 /// handle on the bootstrap server
 pub struct BootstrapManager {
     update_handle: thread::JoinHandle<Result<(), BootstrapError>>,
     // need to preserve the listener handle up to here to prevent it being destroyed
     #[allow(clippy::type_complexity)]
-    _listen_handle: thread::JoinHandle<Result<Result<(), BsConn>, BootstrapError>>,
     main_handle: thread::JoinHandle<Result<(), BootstrapError>>,
```
Comment (Ben-PH) [inline massa-bootstrap/src/server.rs:77]: change to `event_loop_handle`

File: massa-bootstrap/src/server.rs [hunk 7]
```diff
@@ -59,33 +59,56 @@ use white_black_list::*;
...
-    listen_stopper_tx: crossbeam::channel::Sender<()>,
+    listener_stopper: Option<BootstrapListenerStopHandle>,
     update_stopper_tx: crossbeam::channel::Sender<()>,
 }
 
 impl BootstrapManager {
+    /// create a new bootstrap manager, but no means of stopping the listener
+    /// use [`set_listen_stop_handle`] to set the handle
```
Comment (Ben-PH) [inline massa-bootstrap/src/server.rs:84]: create follow-up issue to handle this.

**Review by Ben-PH — APPROVED** [review]
With follow up issues, lgtm. Nice work, @modship

**Comment by modship** [comment]
> With follow up issues, lgtm. Nice work, @modship

Thanks @Ben-PH , u too :v:

**Review by AurelienFT — APPROVED** [review]
Seems ok to me if it has been tested on labnet i'm ok to merge. If you need help to fix conflicts related to network refactoring ping me on discord.

**Pull request merged**